modules/core/04-channel/keeper/upgrade.go

```diff
@@ -0,0 +1,87 @@
+package keeper
+
+import (
+	"strings"
+
+	errorsmod "cosmossdk.io/errors"
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+
+	clienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+	"github.com/cosmos/ibc-go/v7/modules/core/04-channel/types"
+	host "github.com/cosmos/ibc-go/v7/modules/core/24-host"
+)
+
+// ChanUpgradeInit is called by a module to initiate a channel upgrade handshake with
+// a module on another chain.
+func (k Keeper) ChanUpgradeInit(
+	ctx sdk.Context,
+	portID string,
+	channelID string,
+	chanCap *capabilitytypes.Capability,
+	propsedUpgradeChannel types.Channel,
+	counterpartyTimeoutHeight clienttypes.Height,
+	counterpartyTimeoutTimestamp uint64,
+) (uint64, error) {
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		return 0, errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if channel.State != types.OPEN {
+		return 0, errorsmod.Wrapf(types.ErrInvalidChannelState, "expected %s, got %s", types.OPEN, channel.State)
+	}
+
+	if !k.scopedKeeper.AuthenticateCapability(ctx, chanCap, host.ChannelCapabilityPath(portID, channelID)) {
+		return 0, errorsmod.Wrapf(types.ErrChannelCapabilityNotFound, "caller does not own capability for channel, port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	if propsedUpgradeChannel.State != types.INITUPGRADE || propsedUpgradeChannel.Counterparty.PortId != channel.Counterparty.PortId ||
+		propsedUpgradeChannel.Counterparty.ChannelId != channel.Counterparty.ChannelId {
+		return 0, errorsmod.Wrap(types.ErrInvalidChannel, "proposed channel upgrade is invalid")
+	}
+
+	if strings.TrimSpace(propsedUpgradeChannel.Version) == "" {
+		return 0, errorsmod.Wrap(types.ErrInvalidChannelVersion, "channel version must be not be empty")
+	}
+
+	if !channel.Ordering.SubsetOf(propsedUpgradeChannel.Ordering) {
+		return 0, errorsmod.Wrap(types.ErrInvalidChannelOrdering, "channel ordering must be a subset of the new ordering")
+	}
+
+	upgradeSequence, found := k.GetUpgradeSequence(ctx, portID, channelID)
+	if !found {
```

Comment:
technically this could all be replaced with

```go
         upgradeSequence++
         k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
```

but maybe the more explicit version is better

Reply:
https://github.com/cosmos/ibc-go/pull/3331#discussion_r1147334242